xdsl/dialects/x86/ops.py

```diff
@@ -0,0 +1,290 @@
+            _assembly_arg_str(arg)
+            for arg in self.assembly_line_args()
+            if arg is not None
+        )
+        return _assembly_line(instruction_name, arg_str, self.comment)
+
+
+class DoubleOperandInstruction(IRDLOperation, X86Instruction, ABC):
+    """
+    Base class for instructions that take two operands.
+    """
+
+
+class RROperation(Generic[R1InvT, R2InvT], DoubleOperandInstruction):
+    """
+    A base class for x86 operations that have two registers.
+    """
+
+    r1 = operand_def(R1InvT)
+    r2 = operand_def(R2InvT)
+
+    result = result_def(R1InvT)
+
+    def __init__(
+        self,
+        r1: Operation | SSAValue,
+        r2: Operation | SSAValue,
+        *,
+        comment: str | StringAttr | None = None,
+        result: R1InvT,
+    ):
+        if isinstance(comment, str):
+            comment = StringAttr(comment)
+
+        super().__init__(
+            operands=[r1, r2],
+            attributes={
+                "comment": comment,
+            },
+            result_types=[result],
+        )
+
+    def assembly_line_args(self) -> tuple[AssemblyInstructionArg | None, ...]:
+        return self.r1, self.r2
+
+
+@irdl_op_definition
+class AddOp(RROperation[GeneralRegisterType, GeneralRegisterType]):
+    """
+    Adds the registers r1 and r2 and stores the result in r1.
+    x[r1] = x[r1] + x[r2]
+    https://www.felixcloutier.com/x86/add
+    """
+
+    name = "x86.add"
+
+
+# region Assembly printing
+def _append_comment(line: str, comment: StringAttr | None) -> str:
+    if comment is None:
+        return line
+
+    padding = " " * max(0, 48 - len(line))
+
+    return f"{line}{padding} # {comment.data}"
+
+
+def _assembly_arg_str(arg: AssemblyInstructionArg) -> str:
+    if isa(arg, AnyIntegerAttr):
+        return f"{arg.value.data}"
+    elif isinstance(arg, int):
+        return f"{arg}"
+    elif isinstance(arg, str):
+        return arg
+    elif isinstance(arg, GeneralRegisterType):
+        return arg.register_name
+    else:
+        if isinstance(arg.type, GeneralRegisterType):
+            reg = arg.type.register_name
+            return reg
+        else:
+            assert False, f"{arg.type}"
+
+
+def _assembly_line(
+    name: str,
+    arg_str: str,
+    comment: StringAttr | None = None,
+    is_indented: bool = True,
+) -> str:
+    code = "    " if is_indented else ""
+    code += name
+    if arg_str:
+        code += f" {arg_str}"
+    code = _append_comment(code, comment)
+    return code
+
+
+def print_assembly(module: ModuleOp, output: IO[str]) -> None:
+    for op in module.body.walk():
+        assert isinstance(op, X86Op), f"{op}"
+        asm = op.assembly_line()
+        if asm is not None:
+            print(asm, file=output)
+
+
+def x86_code(module: ModuleOp) -> str:
+    stream = StringIO()
+    print_assembly(module, stream)
+    return stream.getvalue()
+
+
+# endregion
+
+
+class GetAnyRegisterOperation(Generic[R1InvT], IRDLOperation, X86Op):
```

Comment:
Is this only used for testing? If so, you should be able to use the `test.op`.
Otherwise, I would advise to have a variadic result_def here, so you can create as many registers in one go as you want.

Reply:
Currently it is only for testing purposes however according to the riscv dialect documentation (from where I got this from) it also has the added functionality of "bridging the RISC-V convention that stores the result of function calls in `a0` and `a1` into SSA form". I haven't gotten to that part yet, but x86 has similar convention in places so perhaps it's better to keep it as it might be useful later down the line?

Reply:
My personal feeling would be to go without this operation for now. It can still be added once it's needed. But that's a personal feeling, I don't mind it being here either.

Reply:
I like how it keeps a certain level of consistency between the assembly dialects, but if the general consensus is to remove it im fine with that as well 👍

Reply:
I agree with Michel, I would propose adding it only when you're sure that there's no alternative